airflow/providers/exasol/operators/exasol.py

```diff
@@ -17,12 +17,15 @@
 # under the License.
 from __future__ import annotations
 
-import warnings
 from typing import Sequence
 
 from airflow.providers.common.sql.operators.sql import SQLExecuteQueryOperator
 
 
+def exasol_fetch_all_handler(cursor) -> list[tuple] | None:
+    return cursor.fetchall()
```

Comment:
I think this function should be in exasol hook rather than the operator?

BTW is this specific to exasol or may applied to other integrations ? if so maybe this needs to be in common.sql ?


I wonder what happens if using `SqlSensor` with exasol?

Reply:
Right. Might be good idea to make it a generic non-fully-dbapi compliant DB (but I am not sure if we have many :))

Reply:
Re: applying to others: Moving it now to common.sql though is an extra complication - because then it means that you can only rely on version >= 1.4.0 (?) for exasol. If there are other reasons to do it, we could make it "common" but if we do not need it now, we can leave it in exasol and move it to common when needed.

There is this nice saying "You have to make something usable first in order to make it reusable".

Reply:
As far as I know at least Presto not implements all mandatory properties, for example [presto-python-client](https://github.com/prestodb/presto-python-client/blob/57b95549650aad5a647a768472bf6909320f1bf3/prestodb/dbapi.py#L37-L40) not implements [paramstyle](https://peps.python.org/pep-0249/#paramstyle) but for this case we manually set this value in [Hook](https://github.com/apache/airflow/blob/8290ade26deba02ca6cf3d8254981b31cf89ee5b/airflow/providers/presto/hooks/presto.py#L83) but technically we could get this value from DB-API compliant libraries.

Reply:
> I wonder what happens if using SqlSensor with exasol?

This will work fine because sensor uses "get_records" method of the hook and does not use handler - and the root cause of the problem was detection of whether the resultSet is returned based on presence of the "description" field. I am doing it differently now in the "exasol-specific" handler.

Reply:
> As far as I know at least Presto not implements all mandatory properties, for example [presto-python-client](https://github.com/prestodb/presto-python-client/blob/57b95549650aad5a647a768472bf6909320f1bf3/prestodb/dbapi.py#L37-L40) not implements [paramstyle](https://peps.python.org/pep-0249/#paramstyle) but for this case we manually set this value in [Hook](https://github.com/apache/airflow/blob/8290ade26deba02ca6cf3d8254981b31cf89ee5b/airflow/providers/presto/hooks/presto.py#L83) but technically we could get this value from DB-API compliant libraries.

I do not think this is a problem - we only really should care about the non-compliance if we are depending on it in our code - otherwise, it's up to the user to know that Presto is not fully db-api-2 compliant. In case of exasol, the root cause was that it does not set the ".description" property which we used to determine if we should fetch results or not, and I believe we have not used `paramstyle` in our "common.sql" code.

Reply:
> Re: applying to others: Moving it now to common.sql though is an extra complication - because then it means that you can only rely on version >= 1.4.0 (?) for exasol. If there are other reasons to do it, we could make it "common" but if we do not need it now, we can leave it in exasol and move it to common when needed.

BTW. This is not even possible now. We needed to implement "exasol-specific" handler. The way how we can check if the result set is present for the ExaStatement is different than for any other handler (`self.result_type == 'resultSet'`) - so this handler is not even a generic one.

Reply:
>  do not think this is a problem - we only really should care about the non-compliance if we are depending on it in our code - otherwise, it's up to the user to know that Presto is not fully db-api-2 compliant. In case of exasol, the root cause was that it does not set the ".description" property which we used to determine if we should fetch results or not, and I believe we have not used paramstyle in our "common.sql" code.

Yep most agree with this with a small concern then we should use all attributes/methods of DB-API 2 first rather than implement our solution because there is always a risk that we reinvent sqlalchemy

If we potentially can use `paramstyle` it would allow us not reinvent the wheel. Right now we by default use  that by DB use `format` (ANSI C printf format codes, e.g. `...WHERE name=%s`) and if DB use different style than we overwrite this argument for specific hook

https://github.com/apache/airflow/blob/3eee33ac8cb74cfbb08bce9090e9c601cf98da44/airflow/providers/common/sql/hooks/sql.py#L126-L127 

Unfortunetly for some reason not all backends implement this attribute such as `presto-python-client` as well as `pyexasol` which also for some reason not implements params for [execute](https://github.com/exasol/pyexasol/blob/master/pyexasol/db2/__init__.py#L37-L38).

Reuse mandatory DB-API 2 parameters and methods would make our life easier in case of if Hook support different backends, for example MySqlHook could use `mysql-connector-python` and `mysqlclient` fortunately both of them use same param style. 

In some time in the future we would in position when we need support both `psycopg2` and `psycopg` (v3) for PostgresHook, but we also lucky with this backends, both of them support `pyformat` (specified in DB-API adapter) and `format`